src/osd/ReplicatedPG.cc

```diff
@@ -7151,11 +7182,17 @@ struct C_Copyfrom : public Context {
   void finish(int r) {
     if (r == -ECANCELED)
       return;
-    pg->lock();
-    if (last_peering_reset == pg->get_last_peering_reset()) {
-      pg->process_copy_chunk(oid, tid, r);
+    if (r == Context::NO_PGLOCK) {
+      if (last_peering_reset == pg->get_last_peering_reset()) {
+	pg->process_copy_chunk(oid, tid, r);
```

Comment:
The value Context::NO_PGLOCK is given only when current thread is ShardedOpWQ holding PG lock

So, I think it'll work correctly